src/main/java/com/google/devtools/build/lib/skyframe/FilesystemValueChecker.java

```diff
@@ -417,6 +422,14 @@ private boolean actionValueIsDirtyWithDirectSystemCalls(ActionExecutionValue act
         try {
           ArtifactFileMetadata fileMetadata =
               ActionMetadataHandler.fileMetadataFromArtifact(file, null, tsgm);
+          FileArtifactValue fileValue = actionValue.getArtifactValue(file);
+          boolean lastSeenRemotely = fileValue != null && fileValue.isRemote();
```

Comment:
When is `fileValue` null here? Is my understanding correct that this is for cases when we don't actually need to call `stat()` to get the metadata of the output of an action?

Reply:
> When is `fileValue` null here?

it's expected to be `null` on clean builds.

> Is my understanding correct that this is for cases when we don't actually need to call `stat()` to get the metadata of the output of an action?

I wasn't aware that this was ever the case in Bazel?

Reply:
Yes. IIRC `fileValue` here is none null only for injected metadata.